src/raftstore/store/peer.rs

```diff
@@ -2007,44 +2010,80 @@ impl RequestInspector for Peer {
 }
 
 #[derive(Debug)]
-pub struct ReadExecutor<'r, 'e, 't> {
-    region: &'r metapb::Region,
-    engine: &'e Arc<DB>,
-    tag: &'t str,
+pub struct ReadExecutor {
+    check_epoch: bool,
+    engine: Arc<DB>,
+    snapshot: Option<SyncSnapshot>,
+    snapshot_time: Option<Timespec>,
+    need_snapshot_time: bool,
 }
 
-impl<'r, 'e, 't> ReadExecutor<'r, 'e, 't> {
-    pub fn new(region: &'r metapb::Region, engine: &'e Arc<DB>, tag: &'t str) -> Self {
+impl ReadExecutor {
+    pub fn new(engine: Arc<DB>, check_epoch: bool, need_snapshot_time: bool) -> Self {
         ReadExecutor {
-            region,
+            check_epoch,
             engine,
-            tag,
+            snapshot: None,
+            snapshot_time: None,
+            need_snapshot_time,
+        }
+    }
+
+    #[inline]
+    pub fn snapshot_time(&mut self) -> Option<Timespec> {
+        self.maybe_update_snapshot();
+        self.snapshot_time
+    }
+
+    #[inline]
+    fn maybe_update_snapshot(&mut self) {
+        if self.snapshot.is_some() {
+            return;
         }
+        // Reading current timespec before snapshot.
+        // The order does not matter in the current implementation,
```

Comment:
What if lease expires right after L2047 and before L2051?

Reply:
Or what if lease becomes suspect?

Reply:
It expires too.